### PR TITLE
first

### PR DESCRIPTION
**Pull request by voloshynamaryna11**

### CODE REVIEW

File: src/main/java/internet/market/Main.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import internet.market.dao.jdbc.ProductDaoJdbcImpl;
+import internet.market.model.Product;
+
+public class Main {
+
+    public static void main(String[] args) {
+        ProductDaoJdbcImpl productDaoJdbc = new ProductDaoJdbcImpl();
+        Product product1 = new Product("mediator", 100);
+        productDaoJdbc.create(product1);
+        System.out.println(productDaoJdbc.get(1L));
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/Main.java]:
```suggestion
        System.out.println(productDaoJdbc.get(product1.getId());
```

Comment (KostiantynVorobiov) [reply src/main/java/internet/market/Main.java]: лучше так

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        String query = "INSERT INTO products (name, price) VALUES (?, ?) ";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection
+                    .prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            statement.setString(1, item.getName());
+            statement.setDouble(2, item.getPrice());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                item.setId(resultSet.getLong("product_id"));
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
                item.setId(resultSet.getLong(1));
```

Comment (KostiantynVorobiov) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: Бодя расказивал на лекции что может выкидивать ошибку

**Review by KostiantynVorobiov — CHANGES_REQUESTED** [review]

File: src/main/java/internet/market/Main.java (hunk 1 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/Main.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 2 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/Main.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package internet.market;
+
+import internet.market.dao.jdbc.ProductDaoJdbcImpl;
+import internet.market.model.Product;
+
+public class Main {
+
+    public static void main(String[] args) {
+        ProductDaoJdbcImpl productDaoJdbc = new ProductDaoJdbcImpl();
```
Comment (boroda4436) [inline src/main/java/internet/market/Main.java]:
```suggestion
        ProductDao productDao = new ProductDaoJdbcImpl();
```

File: src/main/java/internet/market/Main.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        Product product1 = new Product("mediator", 100);
+        productDaoJdbc.create(product1);
+        System.out.println(productDaoJdbc.get(1L));
+        Product product2 = new Product("aF", 234);
+        product2.setId(1L);
```
Comment (boroda4436) [inline src/main/java/internet/market/Main.java]: no. DB should be responsible for id creating

File: src/main/java/internet/market/dao/GenericDao.java [hunk 5]
```diff
@@ -1,16 +1,17 @@
 package internet.market.dao;
 
+import internet.market.exceptions.DataProcessingException;
 import java.util.List;
 import java.util.Optional;
 
 public interface GenericDao<T, K> {
-    T create(T item);
+    T create(T item) throws DataProcessingException;
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/GenericDao.java]: this is not required

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            return item;
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t add this product", e);
+        }
+    }
+
+    @Override
+    public Optional<Product> get(Long id) {
+        String query = "SELECT * FROM products WHERE deleted=0 AND product_id=?";
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: See Sofia's comment in the slack

Comment (boroda4436) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: and deleted = false

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    @Override
+    public List<Product> getAll() {
+        String query = "SELECT * FROM products WHERE deleted=0";
+        List<Product> productList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            while (resultSet.next()) {
+                Long productId = resultSet.getLong("product_id");
+                String productName = resultSet.getString("name");
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: code duplication on line 45

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
+                String productName = resultSet.getString("name");
+                double productPrice = resultSet.getDouble("price");
+                Product product = new Product(productName, productPrice);
+                product.setId(id);
+                return Optional.of(product);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t find product with this id", e);
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Can`t find product with id = " + id, e);
```

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    public Product update(Product item) {
+        String query = "UPDATE products SET name=?, price=? WHERE product_id=?";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, item.getName());
+            statement.setDouble(2, item.getPrice());
+            statement.setLong(3, item.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t add this product", e);
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("We can`t add product", e);
```

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t get all products", e);
+        }
+        return productList;
+    }
+
+    @Override
+    public Product update(Product item) {
+        String query = "UPDATE products SET name=?, price=? WHERE product_id=?";
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java:71]: and deleted = false

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t add this product", e);
+        }
+        return item;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE products SET deleted=1 WHERE product_id=?";
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: See Sofia's comment in the slack

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        boolean deletedUnit;
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            deletedUnit = statement.executeUpdate() > 0;
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            return statement.executeUpdate() > 0;
```

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t add this product", e);
+        }
+        return item;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE products SET deleted=1 WHERE product_id=?";
+        boolean deletedUnit;
```
Comment (boroda4436) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: remember today lesson

File: src/main/java/internet/market/util/ConnectionUtil.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+            Class.forName("com.mysql.cj.jdbc.Driver");
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException("Can't find mysql Driver", e);
+        }
+    }
+
+    public static Connection getConnection() {
+        Properties dbProperties = new Properties();
+        dbProperties.put("user", "root");
+        dbProperties.put("password", "puv34g7hhe4jkopp2!1e$");
```
Comment (boroda4436) [inline src/main/java/internet/market/util/ConnectionUtil.java:20]: never push your real credentials to github

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/internet/market/Main.java (hunk 3 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/Main.java]: done

File: src/main/java/internet/market/dao/GenericDao.java (hunk 5 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/GenericDao.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 6 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 8 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 10 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java:71]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 9 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 11 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 12 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 13 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 7 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/Main.java (hunk 4 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/Main.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+            statement.setString(1, item.getName());
+            statement.setDouble(2, item.getPrice());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                item.setId(resultSet.getLong(1));
+            }
+            return item;
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t add product with id = " + item.getId(), e);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: fix this message according to comments from Bohdan

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    public Product update(Product item) {
+        String query = "UPDATE products SET name=?, price=? WHERE product_id=?";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, item.getName());
+            statement.setDouble(2, item.getPrice());
+            statement.setLong(3, item.getId());
+            statement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can`t add product with id = " + item.getId(), e);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: do we add product here, or we update it ?

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE products SET deleted=true WHERE product_id=?";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            return statement.executeUpdate() > 0;
+        } catch (SQLException e) {
+            throw new DataProcessingException("We can`t delete product with id = " + id, e);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: same with this message

**Review by RomanDubovskyi — COMMENTED** [review]
you didn't fix all exception messages, rest seems to be ok

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 15 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 16 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java (hunk 17 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ProductDaoJdbcImpl.java]: done

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**